src/main/java/expensewebservice/ActionableMessageTokenValidator.java

```diff
@@ -0,0 +1,128 @@
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+package expensewebservice;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.*;
+import java.net.*;
+import java.net.URL;
+import java.text.*;
+import java.util.List;
+import java.util.Objects;
+
+import com.nimbusds.jose.*;
+import com.nimbusds.jose.jwk.source.*;
+import com.nimbusds.jose.proc.*;
+import com.nimbusds.jwt.*;
+import com.nimbusds.jwt.proc.*;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/**
+* Token validator for actionable message.
+*/
+public class ActionableMessageTokenValidator {
+    private static final String O365_APP_ID = "48af08dc-f6d2-435f-b2a7-069abd99c086";
+    private static final String O365_OPENID_METADATA_URL = "https://substrate.office.com/sts/common/.well-known/openid-configuration";
+    private static final String O365_TOKEN_ISSUER = "https://substrate.office.com/sts/";
+
+    /**
+    * Constructor.
+    */
+    public ActionableMessageTokenValidator() {
+    }
+    
+    /**
+    * Validates an actionable message token.
+    * @param token The token issued by Microsoft.
+    * @param targetUrl The target URL expected in the token.
+    * @return ActionableMessageTokenValidationResult The result of the validation.
+    */
+    public ActionableMessageTokenValidationResult validateToken(String token, String targetUrl) throws IOException, ParseException, MalformedURLException, BadJOSEException, JOSEException {
+        OpenIdMetadata config = new OpenIdMetadata(O365_OPENID_METADATA_URL);
+        JWT jwt = JWTParser.parse(token);
+        JWSAlgorithm expectedJWSAlg = JWSAlgorithm.parse(jwt.getHeader().getAlgorithm().getName());
+    
+        JWKSource keySource = new RemoteJWKSet(new URL(config.getJsonWebKeyUrl()));
+        JWSKeySelector keySelector = new JWSVerificationKeySelector(expectedJWSAlg, keySource);
+        
+        ConfigurableJWTProcessor jwtProcessor = new DefaultJWTProcessor();
+        jwtProcessor.setJWSKeySelector(keySelector);
+        
+        SecurityContext ctx = null;
+        JWTClaimsSet claimsSet = jwtProcessor.process(token, ctx);
+        return verifyClaims(claimsSet, targetUrl);
+    }
+
+    /**
+    * Validates a set of claims in a JWT token.
+    * @param claims The claims set to validate.
+    * @param targetUrl The expected URL in the audience claim.
+    * @return ActioanbleMessageTokenValidationResult The result of the validation.
+    */
+    private ActionableMessageTokenValidationResult verifyClaims(JWTClaimsSet claims, String targetUrl) {
+        ActionableMessageTokenValidationResult result = new ActionableMessageTokenValidationResult();
+        
+        try {
+            if (!Objects.equals(claims.getIssuer().toLowerCase(), O365_TOKEN_ISSUER)) {
```

Comment:
Why not O365_TOKEN_ISSUER.equalsIgnoreCase(claims.getIssuer())? (Same for the other similar comparisons.)